### PR TITLE
First prototype of the desktop app

### PR DESCRIPTION
**Pull request by zner0L**
- [x] Components work
- [x] App specific UI
- [x] SMPT support
- [x] Secure keystore support
- [x] Notifications
- [ ] IMAP
- [ ] File storage
- [ ] Global request settings

### CODE REVIEW

File: src/electron/email.ts [hunk 1]
```diff
@@ -1,44 +1,54 @@
...
 
-export type SendMessageOptions = {
-    from: string;
-    to: string;
-    subject: string;
-    text: string;
+export type Credentials = {
+    host: string;
+    port: number;
+    secure: boolean;
```
Comment (zner0L) [inline src/electron/email.ts:17]: Why is this a boolean, when we have three options available?

File: src/electron/email.ts [hunk 2]
```diff
@@ -1,44 +1,54 @@
...
+};
+export type RecreateEmailClientsOptions = { imapCredentials?: Credentials; smtpCredentials?: Credentials };
+export type RecreateEmailClientsReturn = Promise<void>;
+export const recreateEmailClients = async (options: RecreateEmailClientsOptions): RecreateEmailClientsReturn => {
+    if (options.imapCredentials) {
+        clients.imap = new ImapFlow({
+            ...options.imapCredentials,
+            auth: { ...options.imapCredentials.auth, pass: (await getPassword('Datenanfragen.de', 'imap')) || '' },
+        });
+        await clients.imap?.connect().catch(() => undefined);
```
Comment (zner0L) [inline src/electron/email.ts:28]: Why is this necessary here? Isn't this what `verifyConnection` is for?

File: src/app/js/store/settings.ts [hunk 3]
```diff
@@ -2,41 +2,87 @@ import create from 'zustand';
...
-                set(
-                    produce((state) => {
-                        state[smtpSetting] = value;
-                    })
-                ),
+            setEmailAccountSetting: async (setting, value) => {
+                if (setting === 'imapPassword') await window.email.setEmailAccountPassword('imap', value as string);
+                else if (setting === 'smtpPassword')
+                    await window.email.setEmailAccountPassword('smtp', value as string);
+                else if (!['imapUseStartTls', 'smtpUseStartTls'].includes(setting))
```
Comment (zner0L) [inline src/app/js/store/settings.ts:54]: When are these settings stored?

File: src/electron/email.ts (hunk 1 above)
Comment (baltpeter) [reply src/electron/email.ts:17]: In the desktop app, we only have two options. I haven't found a way to get nodemailer and imapflow to use insecure connections (not exactly a great loss…).

File: src/electron/email.ts (hunk 2 above)
Comment (baltpeter) [reply src/electron/email.ts:28]: Not sure what you mean. If we use `ensureConnection()` there, every keypress while entering the credentials will trigger one or two error notifications.

File: src/app/js/store/settings.ts (hunk 3 above)
Comment (baltpeter) [reply src/app/js/store/settings.ts:54]: Never (see above). They are only added to make the TS happy since both mobile and desktop use the same function type for setting a setting.

File: src/electron/email.ts (hunk 2 above)
Comment (zner0L) [reply src/electron/email.ts:28]: I get errors here, because I have not entered any IMAP credentials. I don't understand why, we need to open a connection here when we `ensureConnection` later anyway.

File: src/electron/email.ts (hunk 1 above)
Comment (zner0L) [reply src/electron/email.ts:17]: Ah, I see. There is an option to do this, in nodemailer at least: Setting `secure: false` and `ignoreTLS: true`. If we want to forbid insecure connections, we should at least set  `requireTLS: true` as a default. Any preferences on your side?

File: src/electron/email.ts (hunk 2 above)
Comment (zner0L) [reply src/electron/email.ts:28]: But that means I'll get errors each time I type something in the settings. That seems kinda stupid to me, since we call `ensureConnection` anyway, which is doing the exact same thing.

File: src/electron/email.ts (hunk 1 above)
Comment (baltpeter) [reply src/electron/email.ts:17]: > Ah, I see. There is an option to do this, in nodemailer at least: Setting `secure: false` and `ignoreTLS: true`.

To me, the docs are really not clear on this. But I feel no desire to allow unencrypted connections anyway.

> If we want to forbid insecure connections, we should at least set `requireTLS: true` as a default. Any preferences on your side?

I can't find the default value of `requireTLS`. I feel like I remember that it defaulted to `true`, otherwise I would have set it. But I may be misremembering. I'm fine with setting it either way.

Comment (zner0L) [reply src/electron/email.ts:17]: Not sure how ImapFlow behaves though. The docs are a bit unclear as to whether it still connects if the connection does not even support STARTTLS.

File: src/electron/email.ts (hunk 2 above)
Comment (baltpeter) [reply src/electron/email.ts:28]: Well, currently we _don't_ call `ensureConnection()` for IMAP since there are no IMAP functions yet. I haven't thought about what the best way of doing this for IMAP is. If the IMAP credentials are wrong, we certain don't want to show an error every few minutes. But I guess that's your problem now, so feel free to change as you see fit.

Comment (zner0L) [reply src/electron/email.ts:28]: ok, if that is the case I am fine with merging. I can't approve my own PR though.

**Review by baltpeter — APPROVED** [review]
